doc/design/eventdriven-callbacks.md

```diff
@@ -0,0 +1,247 @@
+ - **robust API** we use this term to describe an API which is hard 
+      to misuse. Especially important in concurrent code since a misuse can lead to race 
+      conditions, Heisenbugs and in general hard to debug problems.
+ - **robust API feature** a feature we only support to increase the robustness 
+      of the API and bugs caused by misuse.
+ - (Reactor pattern)[https://en.wikipedia.org/wiki/Reactor_pattern]
+ - (Condition Variable)[https://en.wikipedia.org/wiki/Monitor_(synchronization)#Condition_variables_2]
+ 
+## Design
+
+### General
+The usage should be similar to the _WaitSet_ with a key difference - it should 
+be **event driven** and not a mixture of event and state driven, depending on
+which event is attached, like in the _WaitSet_.
+
+It should have the following behavior:
+
+ - Whenever an event occurs the corresponding callback should be called **once**
+    immediately.
+ - If an event occurs multiple times before the callback was called, the callback
+    should be called **once**.
+ - If an event occurs while the callback is being executed the callback should be 
+    called again **once**.
+
+The following use cases and behaviors should be implemented.
+
+ - The API and ReactAl should be robust this means that it should be very hard 
+    to use the API in the wrong way. Since ReactAl is working concurrently a 
+    wrong usage could lead to race conditions, extremely hard to debug bug reports 
+    (HeisenBugs) and can be frustrating to the user.
+    We list here features marked with [robust] which are only supported to 
+    increase this kind of the robustness.
+
+ - concurrently: attaching a callback at any time from anywhere. This means in particular
+    - Attaching a callback from within a callback.
+    - [robust] Updating a callback from within the same callback 
+      ```cpp
+      void onSampleReceived2(iox::popo::UntypedSubscriber & subscriber) {}
+
+      void onSampleReceived(iox::popo::UntypedSubscriber & subscriber) {
+        myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED, onSampleReceived2);
+      }
+
+      myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED, onSampleReceived);
+      ```
+ - One can attach at most one callback to a specific event of an object.
+    - Attaching a callback to an event where a callback has been already attached overrides
+      the existing callback with the new one.
+ - concurrently: detach a callback at any time from anywhere. This means in particular
+    - Detaching a callback from within a callback.
+    - [robust] When the detach call returns we guarantee that the callback is never called
+      again. E.g. blocks till the callback is removed, if the callback is concurrently 
+      running it will block until the callback is finished and removed.
+      Exception: If a callback detaches itself it blocks until the callback is removed 
```

Comment:
So the callback will/cannot continue after it is removed? (In some cases it could not properly continue I think, e.g. when depends on the data it relies on).

The semantics are not entirely clear here.